### PR TITLE
WIP: input tracking

### PR DESCRIPTION
**Pull request by stefreak**
**What this PR does / why we need it**:
- Input tracking, to allow excluding inputs in a future PR
- Separate parsing template strings from evaluating template strings by generating an AST in the parser
- Implement lazy evaluation
- implement a proxy so we avoid changing too much existing code that uses templates

Todo:
- Integrate all that with the existing code base (Parsing, validating and processing configs)

**Special notes for your reviewer**:

### CODE REVIEW

File: core/src/template-string/ast.ts [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+      }
+    }
+
+    const addTemplateReferenceInformation = (v: TemplateLeaf, keyPath: ObjectPath) => {
+      return new TemplateLeaf({
+        expr: rawTemplateString,
+        value: v.value,
+        inputs: {
+          // key might be something like ["var", "foo", "bar"]
+          // We also add the keypath to get separate keys for ever
```
Comment (stefreak) [inline core/src/template-string/ast.ts:295]:
```suggestion
          // We also add the keypath to get separate keys for every leaf
```

File: core/src/template-string/ast.ts [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+    loc: Location,
+    public readonly functionName: string,
+    public readonly args: TemplateExpression[]
+  ) {
+    super(loc)
+  }
+
+  override evaluate(evaluateArgs: EvaluateArgs): CollectionOrValue<TemplateLeaf> {
+    const args = this.args.map((arg) => arg.evaluate(evaluateArgs))
+    // TODO: handle inputs correctly
```
Comment (stefreak) [inline core/src/template-string/ast.ts:320]:
```suggestion
```

File: core/src/template-string/ast.ts [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+        }
+
+      case "&&":
+        const leftRes = this.left.evaluate(args)
+        const rightRes = this.right.evaluate(args)
+
+        if (isTruthy(leftRes) && isTruthy(rightRes)) {
+          return mergeInputs(rightRes, leftRes)
+        } else {
+          return mergeInputs(leftRes, rightRes)
```
Comment (stefreak) [inline core/src/template-string/ast.ts:117]: We always need to return the input that was falsy in this case, actually:

```
> 1 && null
null
> null && 2
null
```

Comment (stefreak) [reply core/src/template-string/ast.ts:117]: And we shouldn't evaluate the right if left was falsy. Very important

File: core/src/template-string/functions.ts [hunk 4]
```diff
@@ -454,90 +470,124 @@ export function callHelperFunction({
...
   } catch (error) {
-    const _error = new TemplateStringError({
+    throw new TemplateStringError({
       message: `Error from helper function ${functionName}: ${error}`,
     })
-    return { _error }
+  }
+
+  // We only need to augment inputs for primitive args in case skipInputTrackingForCollectionValues is true.
+  const trackedArgs = args.filter((arg) => {
```
Comment (stefreak) [inline core/src/template-string/functions.ts:535]: the input tracking code should live in ast.ts, as that's where all the other input tracking code lives.

File: core/src/template-string/parser.pegjs [hunk 5]
```diff
@@ -8,22 +8,18 @@
...
-    isPrimitive,
     optionalSuffix,
-    missingKeyExceptionType,
-    passthroughExceptionType,
     resolveNested,
     TemplateStringError,
   } = options
 
+  function filledArray(count, value) {
+    return Array.apply(null, new Array(count))
```
Comment (TimBeyer) [inline core/src/template-string/parser.pegjs:19]: I think `Array.from({ length: count }, () => value)` would be a bit more elegant.

File: core/src/template-string/ast.ts [hunk 6]
```diff
@@ -0,0 +1,640 @@
...
+      message: `{endif} block expression should not end up in the final AST`,
+    })
+  }
+}
+
+export class IfBlockExpression extends TemplateExpression {
+  constructor(
+    loc: Location,
+    public readonly condition: TemplateExpression,
+    public ifTrue: TemplateExpression,
```
Comment (stefreak) [inline core/src/template-string/ast.ts:462]: I think `ifTrue` is also optional actually, for example for this expression it would be undefined: `${if ...}{else}Hello world{endif}`

Would be good to add tests for these cases

File: core/src/template-string/ast.ts [hunk 7]
```diff
@@ -0,0 +1,640 @@
...
+          expr: args.rawTemplateString,
+          value: transformed,
+          inputs: {},
+        }),
+        left,
+        right
+      )
+    }
+
+    return mergeInputs(transformed, left, right)
```
Comment (stefreak) [inline core/src/template-string/ast.ts:276]: No need to merge inputs for concatenating arrays. We likely should skip this step for collections.

File: core/src/template-string/ast.ts [hunk 8]
```diff
@@ -0,0 +1,640 @@
...
+    const inner = this.innerExpression.evaluate(args)
+    const innerValue = isTemplateLeaf(inner) ? inner.value : inner
+
+    if (typeof innerValue !== "string" && typeof innerValue !== "number") {
+      throw new TemplateStringError({
+        message: `Expression in bracket must resolve to a string or number (got ${typeof innerValue}).`,
+      })
+    }
+
+    return new TemplateLeaf({
```
Comment (stefreak) [inline core/src/template-string/ast.ts:541]: This wipes the inputs of the inner expression;
We need to actually:

```suggestion
    return inner
```

File: core/src/template-string/ast.ts [hunk 9]
```diff
@@ -0,0 +1,640 @@
...
+    const conditionResult = this.condition.evaluate({
+      ...args,
+      optional: true,
+    })
+
+    // Get value for left hand side if it's a TemplateValue
+    const conditionalValue = isTemplateLeaf(conditionResult) ? conditionResult.value : conditionResult
+
+    // evaluate ternary expression
+    const evaluationResult = conditionalValue ? this.ifTrue.evaluate(args) : this.ifFalse.evaluate(args)
```
Comment (stefreak) [inline core/src/template-string/ast.ts:635]:
```suggestion
    const evaluationResult = isTruthy(conditionalValue) ? this.ifTrue.evaluate(args) : this.ifFalse.evaluate(args)
```

File: core/src/template-string/parser.pegjs [hunk 10]
```diff
@@ -36,97 +31,226 @@
...
+        case "-":
+          return new ast.SubtractExpression(location(), operator, left, right)
+        case "*":
+          return new ast.MultiplyExpression(location(), operator, left, right)
+        case "/":
+          return new ast.DivideExpression(location(), operator, left, right)
+        case "%":
+          return new ast.ModuloExpression(location(), operator, left, right)
+        default:
+          throw new TemplateStringError({ message: `Unrecognized logical operator: ${operator}` })
```
Comment (stefreak) [inline core/src/template-string/parser.pegjs:68]:
```suggestion
          throw new TemplateStringError({ message: `Unrecognized binary operator: ${operator}` })
```

File: core/src/template-string/parser.pegjs [hunk 11]
```diff
@@ -36,97 +31,226 @@
...
+      } else if (e instanceof ast.EndIfBlockExpression) {
+        if (currentCondition === undefined) {
+          throw new TemplateStringError({ message: "Found ${endif} block without a preceding ${if...} block." })
+        }
+        if (nestingLevel === 0) {
+          currentCondition.ifTrue = buildConditionalTree(...ifTrue)
+          currentCondition.ifFalse = buildConditionalTree(...ifFalse)
+          currentCondition.loc.end = e.loc.end
+          rootExpressions.push(currentCondition)
+          currentCondition = undefined
```
Comment (stefreak) [inline core/src/template-string/parser.pegjs:148]:
```suggestion
          currentCondition = undefined
          ifTrue = []
          ifFalse = []
          encounteredElse = false
```

File: core/src/template-string/parser.pegjs [hunk 12]
```diff
@@ -36,97 +31,226 @@
...
+    if (elements.length === 1 && !(isBlockExpression(elements[0]))) {
+      return elements[0]
+    }
+
+    return buildConditionalTree(...elements)
+  }
+
+TemplateStrings
+  = head:(FormatString)+ tail:TemplateStrings? {
+    // This means we are concatenating strings, and there may be conditional block statements
```
Comment (stefreak) [inline core/src/template-string/parser.pegjs:193]:
```suggestion
```

File: project.garden.yml [hunk 13]
```diff
@@ -21,5 +21,8 @@ providers:
   - name: local-kubernetes
     environments: [local]
 variables:
+  foobar:
+    bar: asd
+    baz: $(local.env.HELLO)
```
Comment (stefreak) [inline project.garden.yml:26]:
```suggestion
```

File: core/src/template-string/template-string.ts [hunk 14]
```diff
@@ -87,187 +61,172 @@ export class TemplateError extends GardenError {
...
+}
+
+function parseTemplateString(string: string, unescape: boolean = false) {
+  const parsed: ast.TemplateExpression = parser.parse(string, {
+    ast,
+    TemplateStringError,
+    // TODO: What is unescape?
+    unescape,
+    escapePrefix,
+    optionalSuffix: "}?",
```
Comment (stefreak) [inline core/src/template-string/template-string.ts:92]: Add deprecation warning for optionalSuffix users.

Comment (stefreak) [reply core/src/template-string/template-string.ts:92]: As a replacement, we could allow optional suffix for identifier and member expressions.

Examples:

```${var.something?}```
```${var.something?.else}```

File: core/src/config/template-contexts/base.ts [hunk 15]
```diff
@@ -186,31 +202,53 @@ export abstract class ConfigContext {
...
+    // Wrap normal data using deepMap
+    else if (isTemplateLeafValue(value)) {
+      result = new TemplateLeaf({
+        expr: undefined,
+        value,
+        inputs: {},
+      })
+    } else {
+      // value is a collection
+      result = deepMap(value, (v) => {
```
Comment (stefreak) [inline core/src/config/template-contexts/base.ts:234]: should be `templatePrimitiveDeepMap`

File: core/src/exceptions.ts [hunk 16]
```diff
@@ -300,14 +303,28 @@ export class CloudApiError extends GardenError {
   }
 }
 
+export class TemplateFunctionCallError extends GardenError {
```
Comment (stefreak) [inline core/src/exceptions.ts:306]: Can be removed again

File: core/src/template-string/lazy.ts [hunk 17]
```diff
@@ -0,0 +1,390 @@
...
+          })
+        }
+      }
+
+      const returnExpression = this.yaml[arrayForEachReturnKey]
+
+      output.push(unwrapLazyValues({ value: returnExpression, context: loopContext, opts }))
+    }
+
+    return output
```
Comment (stefreak) [inline core/src/template-string/lazy.ts:299]: mergeInputs is missing

File: core/src/config/template-contexts/base.ts [hunk 18]
```diff
@@ -163,6 +176,10 @@ export abstract class ConfigContext {
         value = resolveTemplateString({ string: value, context: this._rootContext, contextOpts: opts })
       }
 
+      if (isTemplateLeaf(value) || value instanceof LazyValue) {
+        break
```
Comment (stefreak) [inline core/src/config/template-contexts/base.ts:180]: Break is not the right thing to do here for LazyValue; We need to unwrap the lazy value and continue diving into it, actually.

File: core/src/template-string/inputs.ts [hunk 19]
```diff
@@ -103,32 +102,24 @@ export function mergeInputs(
...
-    }
+  if (containsLazyValues(relevantValues)) {
+    return new MergeInputsLazily(source, result, relevantValues)
   }
 
-  for (const v of relevantValues) {
-    let hasLazyValues = false
-
+  for (const v of relevantValues as CollectionOrValue<TemplateLeaf>[]) {
     deepMap(v, (leaf) => {
```
Comment (stefreak) [inline core/src/template-string/inputs.ts:110]: visit instead of deepMap would be more performant

File: core/src/template-string/lazy.ts [hunk 20]
```diff
@@ -149,21 +162,22 @@ export class WrapContextLookupInputsLazily extends LazyValue {
...
   }
 
-  override evaluate(context: ConfigContext, opts: ContextResolveOpts): CollectionOrValue<TemplateValue> {
+  override evaluateImpl(context: ConfigContext, opts: ContextResolveOpts): CollectionOrValue<TemplateValue> {
     const unwrapped = unwrapLazyValues({ value: this.value, context, opts })
     return deepMap(unwrapped, (v, _k, collectionKeyPath) => {
       // Wrap it lazily
       if (v instanceof LazyValue) {
-        return new WrapContextLookupInputsLazily(this.source, v, this.contextKeyPath, this.expr)
+        return new WrapContextLookupInputsLazily(this.source, v, this.contextKeyPath, this.expr, collectionKeyPath)
```
Comment (stefreak) [inline core/src/template-string/lazy.ts:180]: We need to combine `collectionKeyPath` with `this.collectionKeyPathPrefix` here:

```
        return new WrapContextLookupInputsLazily(this.source, v, this.contextKeyPath, this.expr, [...this.collectionKeyPathPrefix, ...collectionKeyPath])
```

File: core/src/template-string/static-analysis.ts [hunk 21]
```diff
@@ -6,55 +6,58 @@
...
+  return false
 }
 
-export function getContextLookupReferences(
-  value: CollectionOrValue<TemplateValue>
-): ObjectPath[] {
-  const refs: ObjectPath[] = []
+export function containsContextLookupReferences(value: CollectionOrValue<TemplateValue>, path: ObjectPath): boolean {
+  for (const keyPath of getContextLookupReferences(value)) {
+    if (startsWith(`${keyPath.join(".")}.`, `${path.join(".")}.`)) {
```
Comment (stefreak) [inline core/src/template-string/static-analysis.ts:47]: what if keypaths contain dots? Maybe better to compare arrays.

File: core/src/template-string/lazy.ts [hunk 22]
```diff
@@ -0,0 +1,475 @@
...
+  }
+
+  override evaluateImpl(context: ConfigContext, opts: ContextResolveOpts): CollectionOrValue<TemplateValue> {
+    const unwrapped = unwrapLazyValues({ value: this.value, context, opts })
+    const unwrappedRelevantValues = this.relevantValues.map((v) => unwrapLazyValues({ value: v, context, opts }))
+    return deepMap(unwrapped, (v) => {
+      if (v instanceof LazyValue) {
+        return new MergeInputsLazily(this.source, v, unwrappedRelevantValues)
+      }
+      return mergeInputs(this.source, v satisfies TemplateLeaf<TemplateLeafValue>, ...unwrappedRelevantValues)
```
Comment (stefreak) [inline core/src/template-string/lazy.ts:155]: Possible infinite recursion if unwrappedRelevantValues is a collection that contains lazy values?

File: core/src/config/base.ts [hunk 23]
```diff
@@ -266,42 +256,40 @@ export function prepareResource({
...
+    // these are allowed
   } else if (kind === "Module") {
-    spec.path = basePath
     spec.configPath = configFilePath
     delete spec.internal
-    return prepareModuleResource(spec, configFilePath, projectRoot)
+    spec = prepareModuleResource(spec, configFilePath, projectRoot)
   } else if (allowInvalid) {
-    return spec
+    // we shouldn't throw
```
Comment (stefreak) [inline core/src/config/base.ts:273]: Do we really need allowInvalid? Why does it exist?

File: core/src/config/project.ts [hunk 24]
```diff
@@ -335,8 +336,7 @@ export const projectSchema = createSchema({
     ),
     defaultEnvironment: joi
       .environment()
-      .allow("")
-      .default("")
+      .default((parent) => parent.environments[0].name)
```
Comment (stefreak) [inline core/src/config/project.ts:339]:
```suggestion
      .default((parent) => parent.environments?.[0]?.name || "")
```

File: core/src/config/project.ts [hunk 25]
```diff
@@ -587,25 +545,75 @@ export const pickEnvironment = profileAsync(async function _pickEnvironment({
...
+
+  const { name: projectName, variables } = projectConfig.config
+
+  const projectVarfileVars = await loadVarfile({
+    configRoot: projectRoot,
+    path: projectConfig.config.varfile,
+    defaultPath: defaultVarfilePath,
+  })
+
+  const environmentConfigContextVariables = new LayeredContext(
```
Comment (stefreak) [inline core/src/config/project.ts:572]: We need tests for some edge cases here:
e.g. 

```
variables:
  my
    deep:
      structure: foo
```
 
with override 
`variables.other.deep.structure: bar`

File: core/src/config/template-contexts/base.ts [hunk 26]
```diff
@@ -150,17 +156,20 @@ export abstract class ConfigContext {
         if (remainder.length > 0) {
           opts.stack.push(stackEntry)
           const res = value.resolve({ key: remainder, nodePath: nestedNodePath, opts })
-          value = res.resolved
+          value = res.result
           message = res.message
           partial = !!res.partial
+        } else {
+          // TODO: improve error message
+          throw new ConfigurationError({
```
Comment (stefreak) [inline core/src/config/template-contexts/base.ts:164]: Maybe instead of error, it should just result in a plain object with all the values the context contains.

File: core/src/config/template-contexts/base.ts [hunk 27]
```diff
@@ -186,31 +195,86 @@ export abstract class ConfigContext {
...
+    for (const [index, layer] of this._layers.entries()) {
+      const isLastLayer = index === this._layers.length - 1
+
+      res = layer.resolve({
+        key,
+        nodePath,
+        opts: {
+          ...opts,
+          // Throw an error if we can't find the value in the last layer, unless allowPartial is set
+          allowPartial: isLastLayer ? opts.allowPartial : false,
```
Comment (stefreak) [inline core/src/config/template-contexts/base.ts:267]: Unecessary; If we do not find anything, we return `CONTEXT_RESOLVE_KEY_NOT_FOUND` instead of throwing.

File: core/test/unit/src/template-string/validation.ts [hunk 28]
```diff
@@ -0,0 +1,237 @@
...
+import { z } from "zod"
+import { parseTemplateCollection } from "../../../../src/template-string/template-string.js"
+import { expect } from "chai"
+import { GenericContext } from "../../../../src/config/template-contexts/base.js"
+import { GardenConfig } from "../../../../src/template-string/validation.js"
+import Joi from "@hapi/joi"
+
+// In the future we might
+// const varsFromFirstConfig = firstConfig.atPath("var") // can contain lazy values
+// const actionConfigWithVars = actionConfig.merge(firstConfig)
```
Comment (stefreak) [inline core/test/unit/src/template-string/validation.ts:18]:
```suggestion
```

File: core/src/template-string/lazy.ts [hunk 29]
```diff
@@ -0,0 +1,522 @@
...
+      // It would be nice if mergeInputs would allow passing `additionalInputs` directly, without wrapping it in `TemplateLeaf`.
+      new TemplateLeaf({ expr: undefined, value: undefined, inputs: this.additionalInputs })
+    ) as R
+  }
+
+  abstract evaluateImpl(context: ConfigContext, opts: ContextResolveOpts): R
+  abstract visitAll(): TemplateExpressionGenerator
+}
+
+export class OverrideKeyPathLazily extends LazyValue {
```
Comment (stefreak) [inline core/src/template-string/lazy.ts:133]: Potentially not needed?

File: core/src/template-string/lazy.ts [hunk 30]
```diff
@@ -0,0 +1,522 @@
...
+
+  override *visitAll(): TemplateExpressionGenerator {
+    yield* visitAll(this.yaml as CollectionOrValue<TemplateValue>)
+  }
+
+  override evaluateImpl(
+    context: ConfigContext,
+    opts: ContextResolveOpts
+  ): Record<string, CollectionOrValue<TemplateValue>> {
+    // Resolve $merge keys, depth-first, leaves-first
```
Comment (stefreak) [inline core/src/template-string/lazy.ts:447]:
```suggestion
```

**Comment by stefreak** [comment]
The improvements in the template-engine from this PR have made it to main via #6745

Whenever necessary, the input-tracking changes are now a relatively easy exercise to add in main.
